rust/cymbal/src/langs/python.rs

```diff
@@ -0,0 +1,93 @@
+use serde::{Deserialize, Serialize};
+use sha2::{Digest, Sha512};
+
+use crate::frames::{Context, ContextLine, Frame};
+
+#[derive(Debug, Clone, Deserialize, Serialize)]
+pub struct RawPythonFrame {
+    #[serde(rename = "abs_path")]
+    pub path: Option<String>, // Absolute path to the file - unused for now
+    pub context_line: Option<String>, // The line of code the exception came from
+    pub filename: String,             // The relative path of the file the context line is in
+    pub function: String,             // The name of the function the exception came from
+    pub lineno: Option<u32>,          // The line number of the context line
+    pub module: Option<String>,       // The python-import style module name the function is in
+    #[serde(default)]
+    pub pre_context: Vec<String>, // The lines of code before the context line
+    #[serde(default)]
+    pub post_context: Vec<String>, // The lines of code after the context line
+    // Default to false as sometimes not present on library code
+    #[serde(default)]
+    pub in_app: bool, // Whether the frame is in the user's code
+}
+
+impl RawPythonFrame {
+    pub fn frame_id(&self) -> String {
+        // We don't have version info for python frames, so we rely on
+        // the module, function, line number and surrounding context to
+        // uniquely identify a frame, with the intuition being that even
+        // if two frames are from two different library versions, if the
+        // files they're in are sufficiently similar
+        let mut hasher = Sha512::new();
+        self.context_line
+            .as_ref()
+            .inspect(|c| hasher.update(c.as_bytes()));
+        hasher.update(self.filename.as_bytes());
+        hasher.update(self.function.as_bytes());
+        hasher.update(self.lineno.unwrap_or_default().to_be_bytes());
```

Comment:
An aside but as I read through this I realised we could inadvertently cause a change to the generated hash if we were ever to re-order the order we put this stuff into the hasher. Should we consider adding them to an array and sorting that before hashing 🤷 Maybe it's unlikely but it feels like something I'd innocently change

Reply:
Since frame storing is basically just persistent caching, I think it's totally fine to invalidate previous results, all it means is cymbal will do some extra work for a while, but it's not the end of the world.